test/src/blockMonitor/blockchainMachine.test.ts

```diff
@@ -0,0 +1,169 @@
+];
+
+interface ExampleState {
+    someNumber: number;
+}
+
+const initialState = {
+    someNumber: 42
+};
+
+class ExampleReducer implements StateReducer<ExampleState, IBlockStub> {
+    getInitialState(block: IBlockStub) {
+        return initialState;
+    }
+    reduce(prevState: ExampleState, block: IBlockStub) {
+        return {
+            someNumber: prevState.someNumber + block.number
+        };
+    }
+}
+
+class ExampleComponent extends Component<ExampleState, IBlockStub> {
+    public handleChanges(prevState: ExampleState, state: ExampleState): void {}
+}
+
+describe("BlockchainMachine", () => {
+    let reducer: ExampleReducer;
+    let spiedReducer: ExampleReducer;
+    let blockProcessor: BlockProcessor<IBlockStub>;
+    let blockCache: BlockCache<IBlockStub>;
+
+    beforeEach(() => {
+        reducer = new ExampleReducer();
+        spiedReducer = spy(reducer);
+
+        blockCache = new BlockCache<IBlockStub>(100);
+        blocks.forEach(b => blockCache.addBlock(b));
+
+        // Since we only need to process events, we mock the BlockProcessor with an EventEmitter
+        const bp: any = new EventEmitter();
+        bp.blockCache = blockCache;
+        blockProcessor = bp as BlockProcessor<IBlockStub>;
+    });
+
+    it("addComponent throws ApplicationError if already started", async () => {
+        const bm = new BlockchainMachine(blockProcessor);
+        await bm.start();
+
+        expect(() => bm.addComponent(new ExampleComponent(reducer))).to.throw(ApplicationError);
+
+        await bm.stop();
+    });
+
+    it("processNewBlock computes the initial state if the parent is not in cache", async () => {
+        const bm = new BlockchainMachine(blockProcessor);
+        bm.addComponent(new ExampleComponent(reducer));
+        await bm.start();
+
+        blockProcessor.emit(BlockProcessor.NEW_BLOCK_EVENT, blocks[0]);
+
+        verify(spiedReducer.getInitialState(anything())).once();
+        verify(spiedReducer.reduce(anything(), anything())).never();
+
+        await bm.stop();
+    });
+
+    it("processNewBlock computes state with reducer and its parent's initial state if the parent's state is not known", async () => {
+        const bm = new BlockchainMachine(blockProcessor);
+        bm.addComponent(new ExampleComponent(reducer));
+        await bm.start();
+
+        blockProcessor.emit(BlockProcessor.NEW_BLOCK_EVENT, blocks[1]);
+
+        // initializer and reducer should both be called once
+        verify(spiedReducer.getInitialState(anything())).once();
+        verify(spiedReducer.reduce(anything(), anything())).once();
+
+        // Check that the reducer was called on the right data
+        const [state, block] = capture(spiedReducer.reduce).last();
+        expect(state).to.deep.equal(initialState);
+        expect(block).to.deep.equal(blocks[1]);
+
+        await bm.stop();
+    });
+
+    it("processNewBlock computes the state with the reducer if the parent's state is known", async () => {
```

Comment:
Maybe just do one test with multiple components

Reply:
Added in bdaadf3d79c7a1b20f3f516aaec3f7db18f7a278.